### PR TITLE
KUNB2B-117 Create a profile user role use case

### PR DESCRIPTION
**Pull request by davidribeiro-kununu**
- Added dependencies of jms/serializer-bundle, kununu/cqrs-bundle, kununu/doctrine-dbal, kununu/rest-api-bundle, league/tactician-bundle, league/tactician-doctrine and ramsey/uuid
- Create migration to populate roles tables with admin, claimer and owner roles
- Removed roles:test command
- Created POST /user-roles endpoint
- Added AssignRoleToUser command use case
- Added GetRoleByUserAndResource query use case

### CODE REVIEW

File: services/src/UseCase/Query/GetRoleByUserAndResource/Query.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+use Kununu\CQRS\Query\Query as CQRSQuery;
+
+final class Query implements CQRSQuery
+{
+    private $userUuid;
+    private $roleName;
+    private $resourceType;
+    private $resourceUuid;
+
+    public function __construct(string $userUuid, string $roleName, string $resourceType, ?string $resourceUuid)
```
Comment (jcostaventureoak) [inline services/src/UseCase/Query/GetRoleByUserAndResource/Query.php:16]: If the query its called `GetRoleByUserAndResource` why do we need the `$roleName` as a parameter?

File: services/src/Controller/UserRoles/UserRoleController.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        $userUuid = $request->get('user_uuid');
+        $roleName = $request->get('role_name');
+        $resourceType = $request->get('resource_type');
+        $resourceUuid = $request->get('resource_uuid', null);
+        $changedBy = $request->get('changed_by');
+
+        $this->commandBus->handle(new Command($userUuid, $roleName, $resourceType, $resourceUuid, $changedBy));
+
+        return $this->queryBus->handle(new Query($userUuid, $roleName, $resourceType, $resourceUuid));
```
Comment (jcostaventureoak) [inline services/src/Controller/UserRoles/UserRoleController.php:34]: Shouldn't we avoid this extra query to get the user_roles and a read model directly from the handler?

File: services/src/UseCase/Query/GetRoleByUserAndResource/ReadModel/UserRole.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+final class UserRole implements DataProviderResult
+{
+    private $userUuid;
+    private $roleName;
+    private $resourceType;
+    private $resourceUuid;
+
+    public function __construct(string $userUuid, string $roleName, string $resourceType, ?string $resourceUuid)
+    {
+        $this->userUuid = $userUuid;
```
Comment (jcostaventureoak) [inline services/src/UseCase/Query/GetRoleByUserAndResource/ReadModel/UserRole.php:18]: I think you forget the the `id` field. We decide to pass it in the [response](https://jira.xing.hh/browse/KUNB2B-117?focusedCommentId=1619238&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1619238)

**Review by jcostaventureoak — CHANGES_REQUESTED** [review]

File: services/src/Domain/Role/Role.php [hunk 4]
```diff
@@ -6,6 +6,14 @@
 
 class Role
 {
+    public const TYPE_ADMIN = 1;
+    public const TYPE_CLAIMER = 2;
+    public const TYPE_OWNER = 3;
+
+    public const KEY_ADMIN = 'admin';
+    public const KEY_CLAIMER = 'claimer';
+    public const KEY_OWNER = 'owner';
```
[the remark this answers is not in the document]
Comment (davidribeiro-kununu) [reply services/src/Domain/Role/Role.php:15]: Sure, I'll change this

File: services/src/UseCase/Command/AssignRoleToUser/ProfileRoleCreator.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+final class ProfileRoleCreator extends AbstractRoleCreator implements RoleCreator
+{
+    protected function canGiveRole(UserUuid $author, Resource $resourceType): bool
+    {
+        return $this->isUserAdmin($author) || $this->isUserInProfile($author, $resourceType);
+    }
+
+    private function isUserInProfile(UserUuid $author, Resource $resourceType): bool
+    {
+        return (bool) $this->repository->userHasRoleForResource($author, $resourceType);
```
[the remark this answers is not in the document]
Comment (davidribeiro-kununu) [reply services/src/UseCase/Command/AssignRoleToUser/ProfileRoleCreator.php:19]: I've talked with Tamara to clarify this, and for a first iteration any user who has access to a profile can assign a role for other user. In short, claimers can create owners and vice-versa

File: services/src/Controller/UserRoles/UserRoleController.php (hunk 2 above)
Comment (davidribeiro-kununu) [reply services/src/Controller/UserRoles/UserRoleController.php:34]: We can't cause we're using CQRS, so a Command use case doesn't return anything. If we need to retrieve the data, then we have to do a query after.
Eg: https://github.com/kununu/profiles/blob/master/services/src/Controller/Profiles/RefreshProfileByUuidController.php

File: services/src/UseCase/Query/GetRoleByUserAndResource/ReadModel/UserRole.php (hunk 3 above)
Comment (davidribeiro-kununu) [reply services/src/UseCase/Query/GetRoleByUserAndResource/ReadModel/UserRole.php:18]: Sure, I'll add it

File: services/src/UseCase/Query/GetRoleByUserAndResource/Query.php (hunk 1 above)
Comment (davidribeiro-kununu) [reply services/src/UseCase/Query/GetRoleByUserAndResource/Query.php:16]: you're right, i'll change it

File: services/src/Controller/UserRoles/UserRoleController.php (hunk 2 above)
Comment (jcostaventureoak) [reply services/src/Controller/UserRoles/UserRoleController.php]: Ok, thanks!

**Review by jcostaventureoak — APPROVED** [review]
LGTM

**Pull request merged**